### PR TITLE
feat: Implement ZkSync Adapter and SpokePool

### PR DESCRIPTION
**Pull request by nicholaspai**
Implements https://linear.app/uma/issue/ACX-169/write-test-zk-sync-contracts

and adds helper scripts

### CODE REVIEW

File: hardhat.config.ts [hunk 1]
```diff
@@ -32,14 +40,39 @@ const config: HardhatUserConfig = {
...
+      },
+    },
+  },
+  zkSyncDeploy: {
+    zkSyncNetwork: "https://zksync2-testnet.zksync.dev",
+    ethNetwork: "goerli", // Can also be the RPC URL of the network (e.g. `https://goerli.infura.io/v3/<API_KEY>`)
+  },
   networks: {
-    hardhat: { accounts: { accountsBalance: "1000000000000000000000000" } },
+    hardhat: { accounts: { accountsBalance: "1000000000000000000000000" }, zksync: compileZk },
```
Comment (mrice32) [inline hardhat.config.ts:61]: Does this overwrite the normal artifacts or does it output to a different directory?

File: scripts/zkSync.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    outputs: [{ name: "", type: "uint256" }],
+    payable: false,
+    stateMutability: "view",
+    type: "function",
+  },
+];
+
+// zksync block explorer does not have contract verification yet so calling contracts via GUI is impossible. This file
+// contains useful scripts to interact with zksync goerli contracts. To run, run:
+// yarn hardhat run ./scripts/zksync.ts --network zksync-goerli
```
Comment (mrice32) [inline scripts/zkSync.ts:37]: nit: could you put this at the top of the file so it's more visible?

File: scripts/zkSync.ts [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+      amountToReturn: toBN("100000000000000000"), // 0.1
+      chainId: originChainId,
+      refundAmounts: [toBN("100000000000000000")], // -0.1
+      leafId: 0,
+      l2TokenAddress: erc20.address,
+      refundAddresses: ["0x9a8f92a830a5cb89a3816e3d267cb7791c16b04d"],
+    },
+    []
+  );
+  const executeReceipt = await execute.wait();
```
Comment (mrice32) [inline scripts/zkSync.ts:147]: nit: should this be commented out with the other write transactions?

**Review by mrice32 — APPROVED** [review]
LGTM! Just a few minor comments and nits. Nice work!!

File: contracts/ZkSync_SpokePool.sol [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        address _crossDomainAdmin,
+        address _hubPool,
+        address _wethAddress,
+        address timerAddress
+    ) SpokePool(_crossDomainAdmin, _hubPool, _wethAddress, timerAddress) {
+        _setZkBridges(_zkErc20Bridge, _zkEthBridge);
+    }
+
+    modifier onlyFromCrossDomainAdmin() {
+        // Formal msg.sender of L1 --> L2 message will be L1 sender.
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply contracts/ZkSync_SpokePool.sol:53]: I'm confirming with zksync dev support but yes this is my understanding as of now. Also, empirically this authentication has worked for goerli --> zkSync goerli messages

File: contracts/ZkSync_SpokePool.sol [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+/**
+ * @notice ZkSync specific SpokePool, intended to be compiled with `@matterlabs/hardhat-zksync-solc`.
+ */
+contract ZkSync_SpokePool is SpokePool {
+    // On Ethereum, avoiding constructor parameters and putting them into constants reduces some of the gas cost
+    // upon contract deployment. On zkSync the opposite is true: deploying the same bytecode for contracts,
+    // while changing only constructor parameters can lead to substantial fee savings. So, the following params
+    // are all set by passing in constructor params where possible.
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply contracts/ZkSync_SpokePool.sol:21]: Yeah we typically keep these variables mutable but perhaps the comment is misleading. I only included it for now as a consideration but I agree this contract should be rarely re-redeployed so i'll add that note

File: contracts/ZkSync_SpokePool.sol [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+    // Bridge used to withdraw ERC20's to L1: https://github.com/matter-labs/v2-testnet-contracts/blob/3a0651357bb685751c2163e4cc65a240b0f602ef/l2/contracts/bridge/L2ERC20Bridge.sol
+    address public zkErc20Bridge;
+
+    // Bridge used to send ETH to L1: https://github.com/matter-labs/v2-testnet-contracts/blob/3a0651357bb685751c2163e4cc65a240b0f602ef/l2/contracts/bridge/L2ETHBridge.sol
+    address public zkEthBridge;
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply contracts/ZkSync_SpokePool.sol:27]: Great point!

File: deploy/016_deploy_zksync_spokepool.ts [hunk 7]
```diff
@@ -0,0 +1,40 @@
+import { HardhatRuntimeEnvironment } from "hardhat/types/runtime";
+import { Wallet } from "zksync-web3";
+import { Deployer } from "@matterlabs/hardhat-zksync-deploy";
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply deploy/016_deploy_zksync_spokepool.ts:3]: Yeah this isn't the `hardhat-deploy` we're ued to, its a special zksync version

[the remark this answers is not in the document]
Comment (nicholaspai) [reply deploy/016_deploy_zksync_spokepool.ts:3]: `@matterlabs/hardhat-zksync-deploy` pretty much only adds the `deploy-zksync` task

File: deploy/016_deploy_zksync_spokepool.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+  const artifact = await deployer.loadArtifact("ZkSync_SpokePool");
+  const args = [
+    L2_ADDRESS_MAP[chainId].zkErc20Bridge,
+    L2_ADDRESS_MAP[chainId].zkEthBridge,
+    hubPool.address, // Set hub pool as cross domain admin since it delegatecalls the ZkSync_Adapter logic.
+    hubPool.address,
+    L2_ADDRESS_MAP[chainId].l2Weth, // l2Weth
+    "0x0000000000000000000000000000000000000000", // timer
+  ];
+  const contract = await deployer.deploy(artifact, args);
```
[the remark this answers is not in the document]
Comment (nicholaspai) [reply deploy/016_deploy_zksync_spokepool.ts:35]: Yeah i don't think this script has access to `hardhat-deploy` at all but I haven't tried importing `hardhat-deploy` so i'll try that now

[the remark this answers is not in the document]
Comment (nicholaspai) [reply deploy/016_deploy_zksync_spokepool.ts:35]: Ah, so one complication is that the usual `deploy` command doesn't know to look in the `artifacts-zk` folder, instead it looks at the `artifacts` folder. So all of the tutorials use the `deploy-zksync` task

File: hardhat.config.ts (hunk 1 above)
Comment (nicholaspai) [reply hardhat.config.ts:61]: Outputs to `artifacts-zk` and `cache-zk` because the compiler is diff

File: scripts/zkSync.ts (hunk 3 above)
Comment (nicholaspai) [reply scripts/zkSync.ts:147]: good point

**Comment by nicholaspai** [comment]
@mrice32 there are a lot of inconsistencies with the deployment scripts in this PR because we use the zksync tutorial's suggested `hardhat-deploy-zksync` npm package instead of `hardhat-deploy`, but I just found out that the latter might actually support zksync now following this [PR](https://github.com/wighawag/hardhat-deploy/pull/276/files). I'll look into this but its not well documented so i'm not super hopeful.

**Comment by mrice32** [comment]
> @mrice32 there are a lot of inconsistencies with the deployment scripts in this PR because we use the zksync tutorial's suggested `hardhat-deploy-zksync` npm package instead of `hardhat-deploy`, but I just found out that the latter might actually support zksync now following this [PR](https://github.com/wighawag/hardhat-deploy/pull/276/files). ~I'll look into this but its not well documented so i'm not super hopeful.~
> 
> I bumped `hardhat-deploy` to a version that allegedly handles `zksync` and `hardhat deploy` task worked!

Very cool!

File: deploy/016_deploy_zksync_spokepool.ts [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+  const { deployer } = await getNamedAccounts();
+
+  // Grab L1 addresses:
+  const { deployments: l1Deployments } = companionNetworks.l1;
+  const hubPool = await l1Deployments.get("HubPool");
+  console.log(`Using l1 hub pool @ ${hubPool.address}`);
+
+  const chainId = parseInt(await getChainId());
+
+  await deploy("ZkSync_SpokePool", {
```
Comment (mrice32) [inline deploy/016_deploy_zksync_spokepool.ts:19]: 👌

**Review by mrice32 — APPROVED** [review]
LGTM! This looks great. Thanks for looking into the hardhat-deploy issue

**Pull request merged**